text/0648-aspects-priority-ordering.md

```diff
@@ -0,0 +1,525 @@
+ensures that newly created nodes inherit aspects from their parent constructs, even if those nodes are added later in the process. See Appendix for
+Pseudocode for the new `invokeAspects` function.
+
+### Is this a breaking change?
+
+No
+
+### What alternative solutions did you consider?
+
+We have two alternate solutions.
+
+#### 1. `AspectType` enum to Encode Read-only and Mutating Aspects
+
+Currently, there’s no clear distinction between read-only and mutating Aspects. We can fix this by introducing an optional aspectType parameter in
+the new `IAspectV2` interface. This parameter will help categorize aspects as either `MUTATING` or `READONLY`. During invocation, we will execute all
+`MUTATING` aspects first, followed by `READONLY` aspects. Aspects without a specified type will run after the mutating ones but before any readonly ones.
+
+```ts
+export enum AspectType {
+  MUTATING,
+  READONLY,
+}
+
+export interface IAspectV2 {
+  aspectType?: AspectType;
+  visit(node: IConstruct): void;
+}
+```
+
+Then, users can specify the aspectType when writing their Aspect:
+
+```ts
+class MyMutatingAspect implements IAspectV2 {
+  // Explicitly encoding the Aspect type:
+  aspectType = AspectType.MUTATING
+  
+  visit(node: IConstruct) { ... }
+}
+```
+
+Pros:
+
+* Easy to understand and implement; minimal effort required from users.
+* Solves the issue of executing read-only aspects before mutating ones.
+
+Cons:
+
+* Does not give users full control over the order of aspect invocation; does not address scenarios where specific ordering among mutating aspects is
+necessary (e.g., ensuring aspects that create nodes are executed before those that mutate them).
+  * Even if we further broke down the AspectType enum to have 3 options: READONLY, MUTATING-IN-PLACE, and MUTATING-ADD-NODE, we still don’t give customers
+  the full ability to control their aspect invocation ordering.
+* Existing users would need to modify their code to adapt to this new encoding of aspects.
+
+#### 2. Multiple Passes of the Construct Tree with a Stabilization Loop
+
+We can perform multiple passes over the construct tree to ensure that if an Aspect creates a new node, that node will inherit its parent’s Aspects and
+be visited. To prevent infinite recursion, we can set a maximum iteration depth (e.g., 100 passes). If invokeAspects hits this limit while still mutating
+the tree, it will throw an error. View pseudocode for this alternate solution in the Appendix.
+
+Pros:
+
+* Solves the first constraint without requiring changes from the customer. Newly created nodes will automatically inherit and apply their parent’s Aspects.
+
+Cons:
+
+* This solution doesn’t address the third constraint of Aspect ordering. We would still need a way to specify the order for cases where mutating and
+validation Aspects apply to the same node.
+* Validation aspects may still run before mutating ones, causing them to throw Errors and fail when they shouldn’t.
```

Comment:
Depending on the definition of ordering, the order could be controlled on the node itself, giving the guarantee that a validation aspect is executed after a modification aspect on the node itself. However If the validation aspect has to scan through its children sub-tree, then you would have a situation where the modification aspects in the sub-tree was not yet executed.
Is this a relevant use-case? I would expect a validation aspect to be applied only to the node itself and not involve processing its children. If the children require a validation, then you would add it as prioritized aspects to the children, so that you can still control the execution order.